src/pymatgen/io/vasp/outputs.py

```diff
@@ -4262,54 +4262,53 @@ class VaspParseError(ParseError):
 
 
 def get_band_structure_from_vasp_multiple_branches(
-    dir_name: str,
+    dir_name: PathLike,
     efermi: float | None = None,
     projections: bool = False,
 ) -> BandStructureSymmLine | BandStructure | None:
     """Get band structure info from a VASP directory.
 
-    It takes into account that a run can be divided in several branches named
-    "branch_x". If the run has not been divided in branches the method will
-    turn to parsing vasprun.xml directly.
+    It takes into account that a run can be divided in several branches,
+    each inside a directory named "branch_x". If the run has not been
+    divided in branches the function will turn to parse vasprun.xml
+    directly from the selected directory.
 
     Args:
-        dir_name: Directory containing all bandstructure runs.
-        efermi: Efermi for bandstructure.
-        projections: True if you want to get the data on site projections if
-            any. Note that this is sometimes very large
+        dir_name (PathLike): Parent directory containing all bandstructure runs.
+        efermi (float): Fermi level for bandstructure.
+        projections (bool): True if you want to get the data on site
+            projections if any. Note that this is sometimes very large
 
     Returns:
-        A BandStructure Object.
-        None is there's a parsing error.
+        A BandStructure/BandStructureSymmLine Object.
+        None if no vasprun.xml found in given directory and branch directory.
     """
-    # TODO: Add better error handling!!!
-    if os.path.isfile(f"{dir_name}/branch_0"):
-        # Get all branch dir names
+    if os.path.isdir(f"{dir_name}/branch_0"):
+        # Get and sort all branch directories
         branch_dir_names = [os.path.abspath(d) for d in glob(f"{dir_name}/branch_*") if os.path.isdir(d)]
-
-        # Sort by the directory name (e.g, branch_10)
         sorted_branch_dir_names = sorted(branch_dir_names, key=lambda x: int(x.split("_")[-1]))
 
-        # Populate branches with Bandstructure instances
-        branches = []
-        for dname in sorted_branch_dir_names:
-            xml_file = f"{dname}/vasprun.xml"
-            if os.path.isfile(xml_file):
-                run = Vasprun(xml_file, parse_projected_eigen=projections)
-                branches.append(run.get_band_structure(efermi=efermi))
-            else:
```

Comment:
I'm considering deprecating and removing this branch altogether because it's already not doing what this function is intended:
https://github.com/materialsproject/pymatgen/blob/6bb293896945106476d6676cb2b65550d791edb5/src/pymatgen/io/vasp/outputs.py#L4306-L4311

When this function falls into this branch, it might suggest either a wrong directory is selected or the directory is not structured as expected, user should really check the provided directory, instead of we still read a `vasprun.xml` directly from that directory?

What do you think? @janosh

Reply:
probably best to keep as is to avoid a breaking change. never used that function myself so don't know the rationale behind returning

```py
Vasprun(xml_file, parse_projected_eigen=projections).get_band_structure(kpoints_filename=None, efermi=efermi)```

if there are no branch dirs

Reply:
> probably best to keep as is to avoid a breaking change

Yes I totally agree with minimizing breaking change whenever possible, so I'm suggesting adding a deprecation warning to that particular branch instead of changing this behavior right away.

Also this function has several TODOs attached regarding error handling, so I assume change is needed:
https://github.com/materialsproject/pymatgen/blob/6bb293896945106476d6676cb2b65550d791edb5/src/pymatgen/io/vasp/outputs.py#L4285-L4286
https://github.com/materialsproject/pymatgen/blob/6bb293896945106476d6676cb2b65550d791edb5/src/pymatgen/io/vasp/outputs.py#L4301-L4302
https://github.com/materialsproject/pymatgen/blob/6bb293896945106476d6676cb2b65550d791edb5/src/pymatgen/io/vasp/outputs.py#L4307-L4309


The workflow of `get_band_structure_from_vasp_multiple_branches` is now like this:
```
--> check dir_name/branch_*/vasprun.xml  --> Yes --> continue to work on multiple branch bandstructure
                          |
                          |
                          No  --> try to read dir_name/vasprun.xml
                                   I think an exception should be raised here because it's already not
                                  "get band structure from multiple branches" from here
```

Reply:
thanks for clarifying. sorry, i missed the deprecating part in [your "deprecating and removing" message](https://github.com/materialsproject/pymatgen/pull/4061#discussion_r1756634794). i think with a one year deprecation period and no one complaining to keep the current behavior during that time, should be fine to remove it